### PR TITLE
feat!: Stop supporting "protected" keyword

### PR DESCRIPTION
**Pull request by RustanLeino**
This PR removes the `protected` keyword from the language. Instead:

* To restrict access from outside the module, use a `provides` clause in the module's export set.
* If you're trying to add conjuncts to a predicate in a refinement module, see
    - `Test/dafny3/CachedContainer.dfy`,
    - `Test/dafny2/StoreAndRetrieve.dfy`, and
    - `Test/dafny2/MonotonicHeapstate.dfy`

    for examples of how to effectively strengthen predicates in refinement modules.

## Background

The keyword `protected` was introduced in support of refinement modules. In particular, by marking a predicate with `protected`, refinement modules gain the ability to add conjuncts to the body of the predicate. To make this sound while allowing refinement modules to take the place of the modules they are refining, the body of the protected predicate was never visible outside the declaring module.

While this feature made it easy to strengthen predicate definitions, which is often desirable in refinement, it is possible to use other features to accomplish this (see the examples listed above). Since the semantics of these alternative, existing features are more standard, they are easier to understand. And since superimposition refinement is not heavily used in programs anyway, it seems better to remove the unusual feature -- it saves explanations and reduces maintenance work on Dafny.

Because `protected` had the necessary side effect of confining a predicate's body inside the defining module, the `protected` keyword could also be used for this purpose alone, even in modules that are never refined. By deprecating the `protected` keyword, this PR breaks such uses. Luckily, this functionality is supported well by _export sets_. If an export set `provides` the predicate (as opposed to `reveals` the predicate), then the signature of the predicate is made available outside the module, but the predicate's body is not.

Finally, the `protected` modifier could also be applied to a `module`. I don't remember the reason for allowing this in the first place, but the effect of it was to disallow refinements of the module. This (probably completely unused) functionality is removed by this PR as well, and there is no replacement for it.

### CODE REVIEW

**Comment by RustanLeino** [comment]
@tjhance, do you think you can apply the workarounds above to replace your use of `protected`?

**Comment by tjhance** [comment]
> @tjhance, do you think you can apply the workarounds above to replace your use of protected?

yeah, I think so

File: docs/_includes/Programs.md [hunk 1]
```diff
@@ -117,40 +116,33 @@ The `static` modifier is used for class members that that
...
+ field                    | ghost
+ newtype                  | -
+ synonym types            | -
+ iterators                | -
+ method                   | ghost static
+ lemma, colemma, comethod | already-ghost static
+ inductive lemma          | already-ghost static
+ constructor              | -
+ function (non-method)    | already-ghost static
+ function method          | already-non-ghost static
```
Comment (davidcok) [inline docs/_includes/Programs.md:141]: Was it intentional to omit extern in various places on this list?

Comment (RustanLeino) [reply docs/_includes/Programs.md:141]: Yes, but perhaps that was foolish. As I edited the table, I saw these mentions of the `extern` keyword, which is no longer supported, so I figured I'd just remove them while I was at it.

**Review by davidcok — APPROVED** [review]

**Pull request merged**